src/day4.scala

```diff
@@ -0,0 +1,55 @@
+// using scala 3.0.2
+
+package day4
+
+import scala.io.Source
+
+@main def run(): Unit =
+  val input = util.Using.resource(Source.fromFile("input/day4"))(_.mkString)
+  val (part1, part2) = answers(input)
+  println(s"The answer of part1 is $part1")
+  println(s"The answer of part2 is $part2")
+
+case class Board(lines: List[List[Int]]):
+  def mapNumbers(f: Int => Int): Board = Board(lines.map(_.map(f)))
+  def columns: List[List[Int]] = lines.transpose
+
+object Board:
+  def parse(inputBoard: String): Board =
+    val lines = inputBoard.split('\n').toList
+    Board(lines.map(parseLine))
+
+  private val lineParser = raw"\d+".r
+  private def parseLine(inputLine: String): List[Int] =
+    lineParser.findAllIn(inputLine).toList.map(_.toInt)
+
+def answers(input: String): (Int, Int) =
+  val inputSections: List[String] = input.split("\n\n").toList
+  val numbers = inputSections.head.split(',').map(_.toInt)
+
+  val numberToTurn = numbers.zipWithIndex.toMap
+  val turnToNumber = numberToTurn.map((key, value) => (value, key))
+  
+  val boards: List[Board] = 
+    inputSections.tail.map(Board.parse).map(_.mapNumbers(numberToTurn))
+  
+  // for each board, the number of turns until it wins
+  val winningTurns: List[(Board, Int)] =
+    for board <- boards
+    yield
+      val lineMin = board.lines.map(line => line.max).min
+      val  colMin = board.columns.map(col => col.max).min
```

Comment:
That is a very unusual formatting choice. There are two common choices in Scala. Either without any alignment:
```suggestion
      val colMin = board.columns.map(col => col.max).min
```
or with alignment of `=`, but the rest stays aligned on the left, never on the right:
```suggestion
      val colMin  = board.columns.map(col => col.max).min
```